consensus/cometbft/service/block_delay.go

```diff
@@ -0,0 +1,106 @@
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package cometbft
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/berachain/beacon-kit/primitives/encoding/json"
+)
+
+// maxDelayBetweenBlocks is the maximum delay between two consequtive blocks.
+// If the last block time minus the previous block time is greater than
+// maxDelayBetweenBlocks, then we reset `FinalizeBlockResponse.NextBlockDelay`
+// to default.
+//
+// This is needed because the network may stall for a long time and we don't
+// want to rush in new blocks as the network resumes its operation.
+const maxDelayBetweenBlocks = 30 * time.Minute
+
+type blockDelay struct {
+	// - genesis time if height is 0 OR
+	// - block time of the last block if height > 0 and time between blocks is
+	// greater than maxDelayBetweenBlocks.
+	InitialTime time.Time `json:"initial_time"`
+
+	// - InitChainRequest.InitialHeight OR
+	// - last block height if time between blocks is greater than maxDelayBetweenBlocks.
+	InitialHeight int64 `json:"initial_height"`
+
+	// PreviousBlockTime is the time of the previous block.
+	PreviousBlockTime time.Time `json:"previous_block_time"`
+}
+
+// CONTRACT: called only once upon genesis during or after InitChain.
+func blockDelayUponGenesis(genesisTime time.Time, initialHeight int64) *blockDelay {
+	return &blockDelay{
+		InitialTime:       genesisTime,
+		InitialHeight:     initialHeight,
+		PreviousBlockTime: genesisTime,
+	}
+}
+
+func blockDelayFromBytes(
+	bz []byte,
+) *blockDelay {
+	var d blockDelay
+	if err := json.Unmarshal(bz, &d); err != nil {
+		panic(fmt.Errorf("failed to unmarshal blockDelay: %w", err))
+	}
+	return &d
+}
+
+// Next returns the duration to wait before proposing the next block.
+func (d *blockDelay) Next(curBlockTime time.Time, curBlockHeight int64,
+	targetBlockTime time.Duration) time.Duration {
+	// Until `timeout_commit` is removed from the CometBFT config,
+	// `FinalizeBlockResponse.NextBlockDelay` can't be exactly 0. If it's set to
+	// 0, then `timeout_commit` from the config will be used, which is not what
+	// we want since we're trying to control the block time.
+	const noDelay = 1 * time.Microsecond
+
+	// Reset the initial time and height if the time between blocks is greater
+	// than maxDelayBetweenBlocks.
+	if curBlockTime.Sub(d.PreviousBlockTime) > maxDelayBetweenBlocks {
+		d.InitialTime = curBlockTime
+		d.InitialHeight = curBlockHeight - 1
+	}
+
+	t := d.InitialTime.Add(targetBlockTime * time.Duration(curBlockHeight-d.InitialHeight))
+	if curBlockTime.Before(t) {
```

Comment:
Should we use real-time here?

```suggestion
	if now := time.Now(); now.Before(t) {
	   t.Sub(now)
```